temporalio/worker/_workflow_instance.py

```diff
@@ -356,39 +357,38 @@ def activate(
         self._current_thread_id = threading.get_ident()
         activation_err: Optional[Exception] = None
         try:
-            # Split into job sets with patches, then signals + updates, then
-            # non-queries, then queries
-            start_job = None
-            job_sets: List[
-                List[temporalio.bridge.proto.workflow_activation.WorkflowActivationJob]
-            ] = [[], [], [], []]
+            # Apply every job, running the loop afterward
+            no_queries = True
             for job in act.jobs:
-                if job.HasField("notify_has_patch"):
-                    job_sets[0].append(job)
-                elif job.HasField("signal_workflow") or job.HasField("do_update"):
-                    job_sets[1].append(job)
-                elif not job.HasField("query_workflow"):
-                    if job.HasField("initialize_workflow"):
-                        start_job = job.initialize_workflow
-                    job_sets[2].append(job)
-                else:
-                    job_sets[3].append(job)
-
-            if start_job:
-                self._workflow_input = self._make_workflow_input(start_job)
-
-            # Apply every job set, running after each set
-            for index, job_set in enumerate(job_sets):
-                if not job_set:
-                    continue
-                for job in job_set:
-                    # Let errors bubble out of these to the caller to fail the task
-                    self._apply(job)
-
-                # Run one iteration of the loop. We do not allow conditions to
-                # be checked in patch jobs (first index) or query jobs (last
-                # index).
-                self._run_once(check_conditions=index == 1 or index == 2)
+                if job.HasField("initialize_workflow"):
+                    self._workflow_input = self._make_workflow_input(
+                        job.initialize_workflow
+                    )
+                elif job.HasField("query_workflow"):
+                    no_queries = False
+                # Let errors bubble out of these to the caller to fail the task
+                self._apply(job)
+
+            # Conditions are not checked on query activations. Query activations always come without
+            # any other jobs.
+            first_task = False
+            try:
+                self._run_once(check_conditions=no_queries)
+            finally:
+                # Ensure the main workflow function task is initialized after a first run of the
+                # event loop, which might execute before-start signals/updates. This is behind
+                # a finally because if those handlers fail, we need still need the main routine
+                # to be initialized in order to fail tasks properly.
```

Comment:
If it weren't for this bit we would've ended up with net-less code here. A bit too bad, but, I think still nicer than grouping things into batches.